[RNMobile] Adds Clipboard Link Suggestion to Image block and Buttons block

ttahmouch opened the pull request:
## Description
> **Is your feature request related to a problem? Please describe.**
> Whenever the user opens the settings for a block that can be linked to a URL (Image, Button), the URL in the clipboard is pasted automatically in the Link to field. In order to remove that automatically link, they need to tap Link to and then request a removal. This behavior is unexpected for the user and has the potential to expose sensitive information inadvertently posing a security issue.
>
> **Describe the solution you'd like**
> The desired solution would introduce the following changes:
>
> ***1. The `Link to` field will NOT be auto-populated***
> ***2. When the `Link to` screen opens a new option will be added below the input field `From Clipboard`***
> ***3. When the user presses on the `From Clipboard` option the input field is populated with the contents of the clipboard***
> -- @antonis 
[**Original Issue**](https://github.com/wordpress-mobile/gutenberg-mobile/issues/4058)

## How has this been tested?

I smoke tested the change in an `iOS` simulator to see its behavior at runtime. I have ***not*** checked `Android` yet.
I also wrote automated integration tests for the UX behavior.

### Designing the Expectations

We had a brief discussion early on before writing any code to understand how the UX should behave. We attempted to capture those expectations in writing to have requirements to implement against.

```
      GIVEN the Link Settings sheet displays,
      AND the Clipboard has a URL copied,
      THEN the `Link to` field in the Link Settings should be populated with the placeholder text, i.e., `Search or type URL`.

      GIVEN the Link Picker sheet displays,
      AND the Clipboard has a URL copied that is different from the contents of the text input field,
      THEN the `From Clipboard` table cell should be populated with the URL from the Clipboard.

      GIVEN the Link Picker sheet displays with the `From Clipboard` table cell,
      WHEN the Clipboard is cleared or changed to something that is NOT a URL,
      THEN the `From Clipboard` table cell should disappear.

      GIVEN the Link Picker sheet displays,
      AND the contents of the Clipboard are IDENTICAL to the contents of the text input field,
      THEN do NOT display the `From Clipboard` table cell.

      GIVEN the Link Picker sheet displays,
      AND the `From Clipboard` table cell is pressed,
      THEN the `Search or type URL` text input field is populated with the URL from the Clipboard,
      AND the `Add this link` table cell is repopulated with the new URL from the Clipboard.
```

### Designing the Tests

I integration tested the `<ImageEdit/>` and `<ButtonEdit/>` components ***implicitly*** by testing the entire `<Editor/>` component from the surface. I wanted to try and make the testing more "pure" by making it resemble a functional test as closely as possible. By that I mean, I wanted to pretend that I have a screen open, and interact with the screen; not an isolated UI element, i.e., `<ImageEdit/>` or `<ButtonEdit/>`.

+ The integration tests in this PR assert UX expectations for the `<LinkSettings/>` component as a whole which includes the subsequent screen with the `<LinkPicker/>`.
+ The integrations tests assert the following:
```
    + The Link Settings should NOT autofill a URL from the Clipboard without user action.
    + The Link Picker should NOT display the Clipboard Link Suggestion if the Clipboard text is NOT a valid URL.
    + The Link Picker should NOT display the Clipboard Link Suggestion if the Clipboard text is a valid URL, but is also the current link URL.
    + The Link Picker should display the Clipboard Link Suggestion if the Clipboard text is a valid URL, and is NOT the current link URL.
    + The Link Settings should be updated with the Clipboard URL if the user selects the Clipboard Link Suggestion.
```
+ Deliberately structured the test cases in "Given-When-Then (GWT) Gherkin-style."
+ Parameterized the tests to see how the same tests may be leveraged for other block types that have `Link Settings`, e.g., `Embed` and `Button` to avoid repeating the tests for those block types.
+ ~Deliberately removed existing unit tests in this module for the moment in order to make transitioning from `react-test-renderer` to `@testing-library/react-native` easier to manage. They will either be re-added in the near future, or converted to complementary integration tests.~
+ I reverted all the test removals when I was resolving a merge conflict. As a result, I moved all of my tests to a new module in the `link-settings` component, i.e., `packages/components/src/mobile/link-settings/test/edit.native.js`. ***(The module should probably be renamed since it's no longer testing the "Edit Button/Image Block" module specifically.)***

```
$ # Run the integration tests.
$ npm run native test -- packages/components/src/mobile/link-settings/test/edit.native.js
...
 PASS  ../components/src/mobile/link-settings/test/edit.native.js (22.169 s)
  <LinkSettings/> from "core/button"
    ✓ should display the LINK SETTINGS with an EMPTY LINK TO field. (644 ms)
    <LinkPicker/>
      Hide Clipboard Link Suggestion - Invalid URL in Clipboard
        ✓ should display the LINK PICKER with NO FROM CLIPBOARD CELL. (197 ms)
      Hide Clipboard Link Suggestion - Valid and Same URL in Clipboard
        ✓ should display the LINK PICKER with NO FROM CLIPBOARD CELL. (302 ms)
      Show Clipboard Link Suggestion - Valid and Different URL in Clipboard
        ✓ should display the LINK PICKER with the FROM CLIPBOARD CELL populated with the URL from the CLIPBOARD. (172 ms)
      Press Clipboard Link Suggestion
        ✓ should display the LINK SETTINGS with the URL from the CLIPBOARDpopulated in the LINK TO field. (180 ms)
  <LinkSettings/> from "core/image"
    ✓ should display the LINK SETTINGS with an EMPTY LINK TO field. (96 ms)
    <LinkPicker/>
      Hide Clipboard Link Suggestion - Invalid URL in Clipboard
        ✓ should display the LINK PICKER with NO FROM CLIPBOARD CELL. (154 ms)
      Hide Clipboard Link Suggestion - Valid and Same URL in Clipboard
        ✓ should display the LINK PICKER with NO FROM CLIPBOARD CELL. (241 ms)
      Show Clipboard Link Suggestion - Valid and Different URL in Clipboard
        ✓ should display the LINK PICKER with the FROM CLIPBOARD CELL populated with the URL from the CLIPBOARD. (154 ms)
      Press Clipboard Link Suggestion
        ✓ should display the LINK SETTINGS with the URL from the CLIPBOARDpopulated in the LINK TO field. (164 ms)

Test Suites: 1 passed, 1 total
Tests:       10 passed, 10 total
Snapshots:   0 total
Time:        22.623 s
```

### Aside
#### Mocking `Image.getSize` and `Clipboard.[gs]etString` for the whole test runtime.

This will help avoid inadvertently making async I/O operations to underlying Native Modules or HTTP when integration-testing our components to ensure more stability in the test runtime.

##### Image.getSize

+ Mocked `Image.getSize` specifically, i.e., not the whole `Image` module from `react-native`.
+ Rewrote the implementation to simply call the provided success callback synchronously.
+ It returns default values of `0` for `width` and `height` pretending to have retrieved the image URI async using HTTP.

##### Clipboard.[gs]etString

+ Mocked `Clipboard.[gs]etString` which is the entire `Clipboard` module from `react-native`.
+ **Note:** `Clipboard` was deprecated in favor of `@react-native-community/clipboard` and will be removed in the future.
+ Rewrote the implementation to simply do nothing on `setString`.
+ Rewrote the implementation to simply return a `Promise` that resolves to `''` on `getString`.
+ Staging strings returned from the `Clipboard` in future integration tests can be as simple as `Clipboard.getString.mockReturnValueOnce( 'value' )`.

#### Resolving runtime error when running `npm run native test`.

```
    console.error
        Warning: Each child in a list should have a unique "key" prop.

        Check the render method of `StylePreview`. See https://reactjs.org/link/warning-keys for more information.
    ...
          86 |              >
          87 |                      <View style={ [ styles.container, { width: itemWidth } ] }>
        > 88 |                              <View style={ styles.imageWrapper }>
             |                              ^
          89 |                                      { isActive &&
          90 |                                              getOutline( [ styles.outline, innerOutlineStyle ] ) }
          91 |                                      <Image
```

[**Related Issue**](https://github.com/WordPress/gutenberg/issues/35888)
[**Related PR**](https://github.com/WordPress/gutenberg/pull/35476)

## Screenshots
### Before (iOS)
https://user-images.githubusercontent.com/3058263/139108306-02a5486b-2281-49ba-8750-697b420bacc6.mp4
### After (iOS)
~https://user-images.githubusercontent.com/3058263/139108312-56d9b2f4-551d-4c5a-a275-da89160a9aa6.mp4~
#### Button Edit Block (iOS)
https://user-images.githubusercontent.com/3058263/141602320-642f7e64-8fa0-404b-bb2e-a1eaac5e92d7.mp4
#### Image Edit Block (iOS)
https://user-images.githubusercontent.com/3058263/141602321-6c24ba02-4b8d-442d-a5cc-86fff9d5c7c5.mp4
#### GridIcons 📋 SVG (iOS)
![image](https://user-images.githubusercontent.com/3058263/142026564-a10a2110-705d-48a8-849f-11fffafecba0.png)

### After - Button Edit Block (Android)
https://user-images.githubusercontent.com/3058263/142917089-19c673fa-5b58-4ab9-bc64-0612d61a3ce4.mp4
### After - Image Edit Block (Android)
https://user-images.githubusercontent.com/3058263/142917140-140f44c5-4210-48ed-99b3-e8fc3ed87e62.mp4

## Types of changes
**Feature**

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Commit:
TT | 4058 | Resolving runtime error when running `npm run native test`.

```
console.error
    Warning: Each child in a list should have a unique "key" prop.

    Check the render method of `StylePreview`. See https://reactjs.org/link/warning-keys for more information.
...
      86 | 		>
      87 | 			<View style={ [ styles.container, { width: itemWidth } ] }>
    > 88 | 				<View style={ styles.imageWrapper }>
         | 				^
      89 | 					{ isActive &&
      90 | 						getOutline( [ styles.outline, innerOutlineStyle ] ) }
      91 | 					<Image
```

Commit:
TT | 4058 | Mocking `Image.getSize` and `Clipboard.[gs]etString` for the whole test runtime.

This will help avoid inadvertently making async I/O operations to underlying Native Modules
or HTTP when integration-testing our components to ensure more stability in the test runtime.

+ Mocked `Image.getSize` specifically, i.e., not the whole `Image` module from `react-native`.
+ Rewrote the implementation to simply call the provided success callback synchronously.
    + It returns default values of `0` for `width` and `height` pretending to have retrieved the image URI async using HTTP.

+ Mocked `Clipboard.[gs]etString` which is the entire `Clipboard` module from `react-native`.
+ **Note:** `Clipboard` was deprecated in favor of `@react-native-community/clipboard` and will be removed in the future.
+ Rewrote the implementation to simply do nothing on `setString`.
+ Rewrote the implementation to simply return a `Promise` that resolves to `''` on `getString`.
+ Staging strings returned from the `Clipboard` in future integration tests can be as simple as `Clipboard.getString.mockReturnValueOnce( 'value' )`.

Commit:
TT | 4058 | TDDing the removal of auto-population of URLs from the Clipboard into the `Link Settings > Link To`
field when editing an `Image` block type.

+ Deliberately removed existing unit tests in this module for the moment in order to make transitioning
from `react-test-renderer` to `@testing-library/react-native` easier to manage. They will either be
re-added in the near future, or converted to complementary integration tests.

+ Deliberately structured the test cases in "Given-When-Then (GWT) Gherkin-style."

+ Added the first test that necessitates the change for forcing auto-population of URLs from the Clipboard
to NOT take place when editing an `<ImageEdit/>` block.

+ Intend to parameterize the tests to see how the same tests may be leveraged for other block types that
have `Link Settings`, e.g., `Embed` and `Button` to avoid repeating the tests for those block types.

+ Integration tests will eventually be written for various planned user experience scenarios including:

```
  GIVEN the Link Settings sheet displays,
  AND the Clipboard has a URL copied,
  THEN the `Link to` field in the Link Settings should be populated with the placeholder text, i.e., `Search or type URL`.

  GIVEN the Link Picker sheet displays,
  AND the Clipboard has a URL copied that is different from the contents of the text input field,
  THEN the `From Clipboard` table cell should be populated with the URL from the Clipboard.

  GIVEN the Link Picker sheet displays with the `From Clipboard` table cell,
  WHEN the Clipboard is cleared or changed to something that is NOT a URL,
  THEN the `From Clipboard` table cell should disappear.

  GIVEN the Link Picker sheet displays,
  AND the contents of the Clipboard are IDENTICAL to the contents of the text input field,
  THEN do NOT display the `From Clipboard` table cell.

  GIVEN the Link Picker sheet displays,
  AND the `From Clipboard` table cell is pressed,
  THEN the `Search or type URL` text input field is populated with the URL from the Clipboard,
  AND the `Add this link` table cell is repopulated with the new URL from the Clipboard.
```

Commit:
Merge branch 'trunk' into feature/4058-add-link-from-clipboard

ttahmouch commented on packages/block-editor/src/components/block-styles/preview.native.js:
```diff
@@ -67,11 +67,11 @@ function StylePreview( { onPress, isActive, style, url } ) {
 			return (
 				<Animated.View
 					style={ [ outlineStyle, { opacity }, styles[ name ] ] }
-					key={ JSON.stringify( outlineStyle ) }
```
I wasn't sure why we previously were `JSON.stringify`ing the children elements `key`s as there likely isn't a guarantee for uniqueness of the JSON string. However, I didn't set a breakpoint and look at the values. Though, even if they were unique, I'm not sure why we'd want to do something that expensive just to distinguish between the `Animated.View` elements. The `index` may not be the most appropriate way to distinguish them, but at the moment I'm not sure why (and they are now unique and the console error log I received when running the test is gone.)

WDYT?

ttahmouch commented on packages/block-library/src/image/test/edit.native.js:
```diff
@@ -1,65 +1,113 @@
+		registerCoreBlocks();
 	} );
 
-	it( 'renders without crashing', () => {
```
Frankly, I'm questioning the value this test adds besides simply just being a starting point for adding more tests. This would be implicitly asserted by just testing anything regarding the component, right?

WDYT?

mkevins replied:
> I'm questioning the value this test adds besides simply just being a starting point for adding more tests

Not totally sure in this case, but it's possible this was part of an intentionally minimal test suite from an earlier time in the project's history. In the past, e2e tests were not even being run, and iirc, there were concerns about both test flakiness and CI job duration affecting all other Gutenberg PRs.

> This would be implicitly asserted by just testing anything regarding the component, right?

Yes, I think this coverage is included in all other tests.

ttahmouch replied:
This comment became `outdated` when I ended up reverting my deletion of these integration tests. That being said, I appreciate the feedback.

> In the past, e2e tests were not even being run
Would our integration tests still be considered "e2e" tests by our definition if our side-effects were stubbed to not cause actual I/O? I can't remember about these existing tests specifically, but all I/O should be stubbed in my new test contributions.

Commit:
TT | 4058 | Generalized the existing test assertion from the `<ImageEdit/>` block to the `<ButtonsEdit/>` block.

+ Added a utility function called `interactionWithUIElement` to interact with an element under test before assertions. This allows us to pretend that the user is interacting with the UI of the `<Editor/>` as if they are using the mobile app in order to stage a test. Right now it's fairly primitive, and is mostly used for `press` events. It can also be used for `layout` events at the moment, but it would need to be elaborated further to support events like `changeText` or `scroll`.
+ Used `interactionWithUIElement` to generalize a test to assert the same expectations regardless of whether the UX is displaying an `<ImageEdit/>` block or `<ButtonsEdit/>` block. The expected UX the test is asserting should be identical regardless of which block is being used in the `<Editor/>`. The only technical difference is how we'd find the elements in the view tree since they have different accessibility props to identify them, e.g., `getByA11yLabel` for the edit button block, and `getByA11yHint` for the edit image block.

Commit:
Merge branch 'trunk' into feature/4058-add-link-from-clipboard

antonis commented:
Hey @ttahmouch 👋 
probably [this PR](https://github.com/WordPress/gutenberg/pull/34846) might have introduced changes that affect your work. Please check if you should merge from `trunk` and resolve any conflicts 🙇

Commit:
TT | 4058 | Tested scenarios where the `From Clipboard` button in the Link Picker **SHOULD/NOT** display to the user.

+ Tested some naive scenarios where the `From Clipboard` button **WOULD** or **WOULD NOT** display given the `Clipboard` has a **VALID** or **INVALID** URL copied. It is naive because it doesn't account for if the user has already copied the URL suggestion from the `Clipboard` into their Link Settings yet. Presumably, we **SHOULD NOT** suggest the URL if it is already the URL in their Link Settings (which will be a future test assertion).
+ Added a [Clipboard SVG](https://www.iconfinder.com/icons/211649/clipboard_icon) as a placeholder asset for the `From Clipboard` button until a final one is decided. It is MIT licensed.
+ Reimplemented `interactionWithUIElement` to **NOT** be an `async` function to avoid unnecessary warnings, i.e., `Warning: An update to ForwardRef(NavigationContainer) inside a test was not wrapped in act(...).`.

Commit:
TT | 4058 | Moving the integration tests for the `<LinkSettings/>` from the `<ImageEdit/>` Block Library module as they are not specific to the Image Edit Block but also the Button Edit Block.

Commit:
Merge branch 'trunk' into feature/4058-add-link-from-clipboard

Commit:
Merge branch 'trunk' into feature/4058-add-link-from-clipboard

Commit:
TT | 4058 | Tested scenarios for when the Clipboard Link Suggestion is hidden or pressed in the Link Picker.

0. Hiding the Clipboard Link Suggestion.

Previously, the Clipboard Link Suggestion was only being hidden in the Link Picker if the text in the Clipboard was NOT a valid URL.
Now, the Clipboard Link Suggestion will also be hidden in the Link Picker if the text is a valid URL, but is the same URL the user already picked.

1. Pressing the Clipboard Link Suggestion.

Previously, this was unimplemented.
Now, the behavior is identical to selecting a Link Suggestion from the search results in the Link Picker, i.e., it will navigate back to the previous screen and pass it the selected URL to change the Block Attributes.

Commit:
Merge branch 'trunk' into feature/4058-add-link-from-clipboard

ttahmouch commented on packages/block-library/src/button/edit.native.js:
```diff
@@ -171,7 +171,7 @@ function ButtonEdit( props ) {
 			placeholder: __( 'Add URL' ),
 			autoFocus: true,
-			autoFill: true,
+			autoFill: false,
```
**Note:** Don't allow the `URL` from the `Clipboard` to autofill the `Link to` field of the `Button Block`.

ttahmouch commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -134,7 +134,7 @@ function LinkSettings( {
 			valueMask,
 			autoFocus: false,
-			autoFill: true,
+			autoFill: false,
```
**Note:** Don't allow the `URL` from the `Clipboard` to autofill the `Link to` field of the `Image Block`.

ttahmouch commented on packages/icons/src/library/clipboard.js:
```diff
@@ -0,0 +1,20 @@
+/**
```
**Note:** Added a ***placeholder*** SVG for the 📋 until I have a final available to swap. It is `MIT` licensed.

ttahmouch replied:
I updated the SVG to use the `Clipboard` `GridIcon` as @iamthomasbishop suggested. It's shown in videos and screenshots.

ttahmouch commented on packages/icons/src/index.js:
```diff
@@ -38,6 +38,7 @@ export { default as chevronRight } from './library/chevron-right';
 export { default as chevronRightSmall } from './library/chevron-right-small';
 export { default as chevronUp } from './library/chevron-up';
 export { default as classic } from './library/classic';
+export { default as clipboard } from './library/clipboard';
```
**Note:** Exporting the 📋 SVG icon from the entry point of the `icons` package to make it useable elsewhere.

ttahmouch commented on test/native/setup.js:
```diff
@@ -201,3 +212,7 @@ jest.mock( '@wordpress/compose', () => {
 	};
 } );
+
+jest.spyOn( Image, 'getSize' ).mockImplementation( ( url, success ) =>
```
**Note:** I added this because we had a side-effect retrieving the Image size in a `useEffect` when an Image Block was being rendered. This was causing warnings to log to the console because the component state was being modified after the initial render when an Image size had been retrieved asynchronously, and no component being re-rendered as a result of the Image size was being waited for in the test assertions. This is expected as all of the tests in the PR assert things completely unrelated to the Image size. I thought it reasonable to default the implementation to be synchronous, and just return default size values. These size values can be mocked as needed in other related tests that need them to be something other than `0`.

```
packages/components/src/mobile/image/index.native.js

useEffect( () => {
...
		if ( url ) {
			Image.getSize( url, ( imgWidth, imgHeight ) => {
			...
			setImageData( metaData );
...
```

ttahmouch commented on test/native/setup.js:
```diff
@@ -154,6 +154,17 @@ jest.mock( 'react-native-reanimated', () => {
+ *
+ * @see node_modules/react-native/Libraries/Components/Clipboard/Clipboard.js
+ */
+jest.mock( 'react-native/Libraries/Components/Clipboard/Clipboard', () => ( {
```
**Note:** This may become irrelevant depending on the timing of the `Clipboard` module being replaced with the API from `react-native-community` [in a parallel PR](https://github.com/WordPress/gutenberg/pull/36328#discussion_r748746259).

ttahmouch replied:
This will probably need to be updated downstream in the PR regarding the RN upgrade that @dcalhoun is implementing.

ttahmouch commented on packages/components/src/mobile/bottom-sheet/link-suggestion-item-cell.native.js:
```diff
@@ -7,7 +7,7 @@ import { Text, View } from 'react-native';
  */
 import { __ } from '@wordpress/i18n';
-import { globe } from '@wordpress/icons';
+import { clipboard, globe } from '@wordpress/icons';
```
**Note:** I had added an MIT-licensed 📋 SVG as a placeholder in the `@wordpress/icons` module. When @iamthomasbishop made me aware of `GridIcons`, I found the [GridIcons 📋](https://raw.githubusercontent.com/Automattic/gridicons/trunk/svg-min/gridicons-clipboard.svg) and swapped the SVGs. This will be published shortly.

ttahmouch replied:
This was published as noted.

ttahmouch commented on packages/components/src/mobile/bottom-sheet/link-suggestion-item-cell.native.js:
```diff
@@ -36,7 +39,7 @@ const getSummaryForType = ( type ) => {
 
 // we use some Cell styles here with a column flex-direction
-function LinkSuggestionItemCell( { suggestion, onLinkPicked } ) {
+function LinkSuggestionItemCell( { suggestion, onLinkPicked, ...props } ) {
```
**Note:** Spreading props here was simply because I would like to cascade `accessibility`-related props to the children elements. I thought about just making them first-class props in the interface of the component as well. I wasn't sure it'd be worth the verbosity.

WDYT?

ttahmouch replied:
> I am not a big fan of this pattern {...props} mostly since it makes it harder to find the props that are being passed down. Do we need to pass down all the props? Why are we doing this in this case?

👆, @enejb

ttahmouch replied:
I added this to my `TODO` list for a future PR.

Commit:
TT | 4058 | Cleaning up comments and console logs in the test module.

Commit:
TT | 4058 | Replacing the placeholder Clipboard SVG with one from GridIcons.

ttahmouch commented on packages/components/src/mobile/gridicons/index.native.js:
```diff
@@ -9,6 +9,9 @@ const fromPathData24x24 = ( pathData ) => (
 );
 
+export const clipboard = fromPathData24x24(
+	'M16 18H8v-2h8v2zm0-6H8v2h8v-2zm2-9h-2v2h2v15H6V5h2V3H6a2 2 0 00-2 2v15a2 2 0 002 2h12a2 2 0 002-2V5a2 2 0 00-2-2zm-4 2V4a2 2 0 10-4 0v1a2 2 0 00-2 2v1h8V7a2 2 0 00-2-2z'
```
**Note:** [The 📋 GridIcon SVG.](https://raw.githubusercontent.com/Automattic/gridicons/trunk/svg-min/gridicons-clipboard.svg)

ttahmouch commented on packages/components/src/mobile/link-picker/index.native.js:
```diff
@@ -47,12 +47,20 @@ export const createDirectEntry = ( value ) => {
 	onCancel: cancel,
 } ) => {
-	const [ value, setValue ] = useState( initialValue );
+	const [ { value, clipboardUrl }, setValue ] = useState( {
```
**Note:** I consolidated the component state into an `object` to avoid needing to `setState` twice (and inadvertently cause more than one needed render). Unfortunately, this made the diff larger than I'd have liked for the module.

ttahmouch commented on packages/components/src/mobile/link-picker/index.native.js:
```diff
@@ -115,6 +133,24 @@ export const LinkPicker = ( {
 					) }
 				</BottomSheet.Cell>
+				{ !! clipboardUrl && clipboardUrl !== value && (
+					<BottomSheet.LinkSuggestionItemCell
```
**Note:** I'm still unsure if it makes sense to encapsulate this element into `<LinkPickerResults/>` below. There is just slightly more margins between the `From Clipboard` suggestion and the search result suggestions because it's not in the same list. It's not easily noticeable because there aren't dividing lines between the suggested items.

WDYT?

![image](https://user-images.githubusercontent.com/3058263/142029521-07f5722b-a5b3-46f0-9927-f2040f76e24c.png)

ttahmouch replied:
This is on my `TODO` list.

Commit:
TT | 4058 | Updated the `accessibilityLabel` of the Clipboard `<LinkSuggestionItemCell/>` to be localized. The tests still default to the `English` translation.

ttahmouch commented on packages/components/src/mobile/link-picker/index.native.js:
```diff
@@ -1,16 +1,16 @@
+import { __, sprintf } from '@wordpress/i18n';
 import { BottomSheet, Icon } from '@wordpress/components';
-import { getProtocol, prependHTTP } from '@wordpress/url';
+import { getProtocol, isURL, prependHTTP } from '@wordpress/url';
```
**Note:** This appeared to be the "standard" URL validator we used in other areas. I leveraged it here as well.

ttahmouch commented:
### TODO
+ [ ] Encapsulate `From Clipboard` into `<LinkPickerResults/>`?
+ [ ] Rename test module `packages/components/src/mobile/link-settings/test/edit.native.js` to `index.native.js`?
+ [ ] Rewrite the `default` localized text for the accessibility label of the `From Clipboard` link suggestion, i.e., `__( 'Copy URL from the clipboard, %s' )`?
+ [x] Confirm functionality on `Android` with a smoke test.
+ [ ] Migrate to `@react-native-clipboard/clipboard` depending on timing?
+ [ ] Modify `...props` spreading to be explicit regarding the `accessibility` props that it is cascading to children elements.
+ [ ] Implement dynamic link suggestions when updating your clipboard and returning Mobile Gutenberg to the foreground.
+ [ ] Remove redundant `jest.mock( 'react-native/Libraries/Components/Clipboard/Clipboard')`.
+ [ ] "Quick question from a translator. In the new strings for Android, there is this string: Copy URL from the clipboard, %s  I would like to clarify: Does 'Copy' here actually mean 'Paste'?"

ttahmouch commented on packages/components/src/mobile/link-picker/index.native.js:
```diff
@@ -1,16 +1,16 @@
  * External dependencies
  */
-import { SafeAreaView, TouchableOpacity, View } from 'react-native';
+import { Clipboard, SafeAreaView, TouchableOpacity, View } from 'react-native';
```
**Note:** I just realized that I may want to change this to be importing from `react-native-community` depending on the timing with @dcalhoun 's [PR for upgrading RN and migrating Clipboard](https://github.com/WordPress/gutenberg/pull/36328).

ttahmouch replied:
This will likely need to be updated downstream to change this to be importing from `react-native-community` in @dcalhoun 's [PR for upgrading RN and migrating Clipboard](https://github.com/WordPress/gutenberg/pull/36328).

antonis reviewed (commented):
Great work @ttahmouch 👍 

I tested the functionality using metro on Android and everything works as expected for me! The code also LGTM on a first pass and I have no comments at this point 🎉 
The added tests also look good. Thank you for taking the effort to add those 🙇 

<details>
<summary>I have a suggestion regarding the behavior when switch back to the app with new contents in the clipboard. In that case the suggestion is not updated. I don't consider this a major issue and I think that we should handle this in a separate PR (if catching this case is feasible).</summary>

https://user-images.githubusercontent.com/304044/142418477-326b8685-e9c2-487e-ba9a-8c9957d4d81f.mp4
</details>

Since the added functionality is sufficient I would suggest to wrap-up this PR as is without extending the functionality/scope at this point (since the PR is already 500 lines) and marking it ready for review. We can then follow up by opening separate issues/PRs for any functionality or code improvements. Wdyt?

enejb commented on packages/components/src/mobile/bottom-sheet/link-suggestion-item-cell.native.js:
```diff
@@ -58,6 +61,7 @@ function LinkSuggestionItemCell( { suggestion, onLinkPicked } ) {
 
 	return (
 		<Cell
+			{ ...props }
```
I am not a big fan of this pattern {...props} mostly since it makes it harder to find the props that are being passed down. 
Do we need to pass down all the props? Why are we doing this in this case?

enejb replied:
I realize we probably do this in other parts of the code base and this might just be my preference :)

ttahmouch replied:
It's not my personal preference either. I just also don't feel particularly strongly about either direction.

ttahmouch replied:
I will resolve this in a smaller future PR. I appreciate your help releasing the current changes, @enejb .

ttahmouch commented:
Apologies for the delay, @antonis . I started replying earlier, and then got distracted.

> Great work @ttahmouch 👍
>
> I tested the functionality using metro on Android and everything works as expected for me! The code also LGTM on a first pass and I have no comments at this point 🎉
> The added tests also look good. Thank you for taking the effort to add those 🙇

Thank you for taking the time to thoroughly review the branch. I appreciate you smoke testing the code on `Android`. That's very helpful.

> I have a suggestion regarding the behavior when switch back to the app with new contents in the clipboard. In that case the suggestion is not updated. I don't consider this a major issue and I think that we should handle this in a separate PR (if catching this case is feasible).

That functionality crossed my mind during development. I just never ended up circling back around to it.

> Since the added functionality is sufficient I would suggest to wrap-up this PR as is without extending the functionality/scope at this point (since the PR is already 500 lines) and marking it ready for review. We can then follow up by opening separate issues/PRs for any functionality or code improvements. Wdyt?

The PR is rather large now, yes. Granted, `432` of those lines are just the test module.

However, I agree. If we were to add this additional enhancement, there's no reason that should block this PR from moving forward. Being able to change the contents of the `Clipboard` suggestion when the `BottomSheet` comes into focus from the background with a `useFocusEffect` or similar is a nice to have, but not MVP, in my opinion. That being said, I don't think the complexity/scope of the changes would be high since we appear to leverage `react-navigation`. Though I'd spike it to be sure.

Does that make sense?

ttahmouch commented:
I just realized that GitHub is dumb, and all the comments I had left when I reviewed my own code the other day were marked as `Pending`. So I don't think they were visible to you when you reviewed the PR, @enejb and @antonis , but I'm unsure.

enejb reviewed (approved):
I tested this as well and I think it's ready to merge. 

We can merge this as it is and address and further changes in other prs.

iamthomasbishop commented:
@ttahmouch Thanks for cranking this out, looks great! I just noticed one tiny thing. I thought it might just be a weird artifact in one of the videos above, but I'm seeing the same thing in all of them. There is a very small (~6-8px) diff in spacing, which results in a "jump" when the cell switches between "from clipboard" and a regular cell.

<a href="https://user-images.githubusercontent.com/1200432/142941375-927a97ab-491c-46f6-946d-0b4cb8d61bff.png"><img src="https://user-images.githubusercontent.com/1200432/142941375-927a97ab-491c-46f6-946d-0b4cb8d61bff.png" width="852"></a>

My guess is that there is supposed to be a bit more padding at the top or above the cell, which would bring the new "from clipboard" cell (on the left in the example above) down to align with the other cell (on the left right in the example above), but I'm not 100% sure.

ttahmouch commented:
> @ttahmouch Thanks for cranking this out, looks great! I just noticed one tiny thing. I thought it might just be a weird artifact in one of the videos above, but I'm seeing the same thing in all of them. There is a very small (~6-8px) diff in spacing, which results in a "jump" when the cell switches between "from clipboard" and a regular cell.
> 
> <img alt="" width="852" src="https://user-images.githubusercontent.com/1200432/142941375-927a97ab-491c-46f6-946d-0b4cb8d61bff.png">
> 
> My guess is that there is supposed to be a bit more padding at the top or above the cell, which would bring the new "from clipboard" cell (on the left in the example above) down to align with the other cell (on the left right in the example above), but I'm not 100% sure.

I think this is a related issue to the Clipboard Link Suggestion not being in the same list as the Search Result Link Suggestions that I noted in this PR. It's been on my TODO list to move the Clipboard Link Suggestion to technically be in the same list, and hopefully that should correct all margin-related issues. Thanks for letting me know, @iamthomasbishop .

fluiddot commented on test/native/setup.js:
```diff
@@ -154,6 +154,17 @@ jest.mock( 'react-native-reanimated', () => {
+ *
+ * @see node_modules/react-native/Libraries/Components/Clipboard/Clipboard.js
+ */
+jest.mock( 'react-native/Libraries/Components/Clipboard/Clipboard', () => ( {
```
@ttahmouch Looks like the Clipboard library is also mocked line 189, I recall I also had to add it for other integration tests. I don't think this would cause any conflict but might be confussing having duplicated code, wdyt?

ttahmouch replied:
Thanks for letting me know, @fluiddot . I think I remember you adding it for work related to the Embed Block that you and Joel implemented.

I think this may have slipped by when I merged the `trunk` branch downstream into my feature branch to keep it up-to-date before release. I can remove the redundancy.

fluiddot replied:
No worries, it happened to me several times 😅 . I wish Git could detect these redundancy cases although they are not trivial at all to infer.

fluiddot commented on packages/components/src/mobile/link-settings/test/edit.native.js:
```diff
@@ -0,0 +1,432 @@
+				</div>
+				<!-- /wp:button -->
+			`,
+			toJSON: () => 'core/button',
```
@ttahmouch I'm wondering if we need this property as I don't see being exposed and used in the function at line 58, should we remove it?

ttahmouch replied:
It is actually used at [line 58](https://github.com/ttahmouch/gutenberg/blob/40f5004851908960a835a013086b56f56cb25cb4/packages/components/src/mobile/link-settings/test/edit.native.js#L58) just not in the way you'd think.

> If the value has a toJSON() method, it's responsible to define what data will be serialized.
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify

`'<LinkSettings/> from %j'` is serialized with the test arguments `toJSON` to make it clear in the console logs that this test suite iteration is for a different block under test, i.e., `core/button` then `core/image`.

I can remove it, but I felt it to be useful. WDYT?

fluiddot commented on packages/components/src/mobile/link-settings/test/edit.native.js:
```diff
@@ -0,0 +1,432 @@
+		const subject = await initializeEditor( { initialHtml } );
+		Clipboard.getString.mockReturnValue( url );
+
+		// Act
+		try {
+			const block = await waitFor( () =>
+				subject.getByA11yLabel(
+					type === 'core/image' ? /Image Block/ : /Button Block/
+				)
+			);
+			fireEvent.press( block );
+			fireEvent.press( block );
+			fireEvent.press(
+				await waitFor( () => subject.getByA11yLabel( 'Open Settings' ) )
+			);
+		} catch ( error ) {
+			done.fail( error );
+		}
+
+		// Assert
+		try {
+			await waitFor( () =>
+				subject.getByA11yLabel(
+					`Link to, ${
+						type === 'core/image' ? 'None' : 'Search or type URL'
+					}`
+				)
+			);
+			done();
+		} catch ( error ) {
+			done.fail( expectation );
+		}
```
@ttahmouch I'm out of curiosity about the use of `try-catch` in these tests, in case we could replicate it on future tests. I understand that the purpose is to control the message displayed when the test fails, right?

ttahmouch replied:
That's correct. I didn't feel that the default error message is particularly expressive or helpful for the final test expectation. Though I considered just decorating the original `{Error}` instance emitted by `react-native-testing-library`. I can still consider changing it to log the original error as well if you think it would be helpful.

WDYT?

fluiddot replied:
No, I think it's great providing more accurate error messages, although it shouldn't be the common case as tests should succeed 😅 . Checking the Action section, I'm wondering if it's necessary the first `try-catch` there because if any error is raised the test would fail in any case before entering the Assert section, right?.

dcalhoun commented on packages/components/src/mobile/link-picker/index.native.js:
```diff
@@ -79,6 +87,14 @@ export const LinkPicker = ( {
 	);
 
+	useEffect( () => {
+		getURLFromClipboard()
+			.then( ( url ) => setValue( { value, clipboardUrl: url } ) )
+			.catch( () => setValue( { value, clipboardUrl: '' } ) );
```
In order to avoid React errors regarding setting state on an unmounted component, we may want to avoid updating the state with a clean up function for the Hook. WDYT? 

```js
useEffect(() => {
  let isCurrent = true;
  getURLFromClipboard()
    .then((url) => {
      if (isCurrent) {
        setValue({ value, clipboardUrl: url });
      }
    })
    .catch(() => {
      if (isCurrent) {
        setValue({ value, clipboardUrl: "" });
      }
    });
  return () => {
    isCurrent = false;
  };
}, []);
```

dcalhoun reviewed (commented):
Nice work! Happy to see this feature landed. 🥳
